### ISSUE TITLE
Error when saving pages using Django 2.0

### ISSUE THREAD

**Issue by simenheg** [stephenmcd/mezzanine#1899]
Using Django 2.0.9 and Mezzanine d437b821, attempting to save a page in the admin interface yields an error along the following lines:

```
TypeError at /nb/admin/pages/richtextpage/1/change/
Direct assignment to the reverse side of a related set is prohibited. Use +.set() instead.
[...]
```

The error traces back to `mezzanine.generic.fields.KeywordsField.save_form_data`, indicating a problem with how keywords are being saved.

**Comment by stephenmcd** [comment]
Also discussed here: https://groups.google.com/forum/#!topic/mezzanine-users/Vtm0N7497Es

**Comment by mfschmidt** [comment]
Thank you for opening the issue, simenheg.
I've replicated this on debian8 and ubuntu18, and with mariadb and postgresql. It also occurs with blogs and richtextpages, but not site settings fields.

**Comment by simenheg** [comment]
I'm not sure that I'm fully understanding the issue, but maybe a simple solution could be to remove the super call from `KeywordsField.save_form_data`:

```diff
diff --git a/mezzanine/generic/fields.py b/mezzanine/generic/fields.py
index d27e4840..57fe4bff 100644
--- a/mezzanine/generic/fields.py
+++ b/mezzanine/generic/fields.py
@@ -203,7 +203,6 @@ class KeywordsField(BaseGenericRelation):
             data = [related_manager.create(keyword_id=i) for i in new_ids]
         # Remove keywords that are no longer assigned to anything.
         Keyword.objects.delete_unused(removed_ids)
-        super(KeywordsField, self).save_form_data(instance, data)
 
     def contribute_to_class(self, cls, name):
         """
```

As far as I can tell, it may not be needed, since the above code is already doing the saving work? The super call ends up in Django's `Field.save_form_data`, which is causing the trouble, since direct assignment to related sets aren't allowed anymore:

```python
def save_form_data(self, instance, data):
    setattr(instance, self.name, data)
```

With this change, saving pages (and other things with keywords related to them) seems to work again in a few simple test cases I've tried.

**Comment by mfschmidt** [comment]
I actually commented out the 'super(...' line this morning, expecting not to have keyword functionality and just to get to the next error. Not only did everything seem to work OK, the keywords continued to save, too. I created numerous blog posts, rich text pages, comments, keywords, etc. without issue. I'm not comfortable enough with my own experience and understanding of the mezzanine and django codebase to suggest the edit for production, but it seems fine here.
From reading many peripheral error reports, based on the django error but not necessarily mezzanine, the problem may have to do with saving keywords to the wrong side of a ReverseManyToOne relationship in the django.db layer. And using the set() function handles it, while the = operator (or setattr?) does not. But I still haven't figured out where the '+' comes from, and don't have strong confidence in my current assessment.
Still looking...

**Comment by eht16** [comment]
After reading https://github.com/django/django/commit/9c5e272860c076736237d4b280c7c922c46ba273, I think `django.db.models.fields.Field:save_form_data` should have been adjusted as well, like:
```python
    getattr(instance, self.name).set(data)
```
I'm not yet sure if this is a bug in Django or desired behavior as I still didn't understand all of the details. Especially since keyword handling in Mezzanine seams to work without the `super` call.

Anyway, assuming it is not a bug in Django, I think the solution is the following:
```diff
--- a/fields.py	2019-01-26 12:51:37.099983174 +0100
+++ b/fields.py	2019-01-26 12:53:15.287986420 +0100
@@ -203,7 +203,8 @@
             data = [related_manager.create(keyword_id=i) for i in new_ids]
         # Remove keywords that are no longer assigned to anything.
         Keyword.objects.delete_unused(removed_ids)
-        super(KeywordsField, self).save_form_data(instance, data)
+        # set data in the Django base class
+        getattr(instance, self.name).set(data)
 
     def contribute_to_class(self, cls, name):
         """
```

It would be awesome if we could sort this out as this is probably the last breaking bit ttowards Django 2.0.

**Comment by simenheg** [comment]
@shangzhikeji and @eht16, have you gained any experience by running your respective branches of Mezzanine which contain the assumed fix above?

It seems to me too that Mezzanine is very close to Django 2.0 compatibility, and that this might indeed be the last tiny hurdle.

Maybe @stephenmcd has some input?

**Comment by eht16** [comment]
My workaround works for me.
So far I didn't notice any drawbacks although I must admit I don't use keywords that intensive. But basic keyword adding, changing and removing works.

Anyway, since my understanding of the whole context is still not better than described above, it would probably be good if @stephenmcd and/or anyone else with more experience could have a deeper look.
Until then, I'd consider my changes rather as workaround than a solution.

**Comment by jerivas** [comment]
#1956 has been merged with complete support for Django 2.2
